[Daniel Lim Wei En] iP

DanielLimWeiEn opened the pull request:
# DukePro

> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,
* text-based
* easy to learn
* ~~FAST~~ _SUPER_ FAST to use

All you need to do is,
1. download it from [here](https://github.com/DanielLimWeiEn/ip/releases/tag/v0.1).
1. double-click it.
1. add your tasks
1. let it manage your tasks for you 😉 

And it is **FREE**!

Features:
* [x] Managing tasks
* [ ] Managing deadlines (coming soon)
* [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method.
```java
public class Duke {

    // ... Code Omitted for Brevity ...

    public static void main(String[] args) {
        new Duke("data/tasks.txt").run();
    }
}
```

Current To Do:
* [x] Add a GUI using JavaFX, increment Level-10
* [ ] Use checkstyle, increment A-CheckStyle
* [ ] Use varargs, increment A-Varargs
* [ ] Add JUnit Tests for more methods
* [ ] Add Exception Handling for more exceptions

Commit:
Add Gradle support

DanielLimWeiEn committed:
Implement a skeletal version of Duke that starts by greeting the user and simply echos the commands entered by the user, and exits when the user types bye.

DanielLimWeiEn committed:
Add the ability to store whatever text entered by the user and display them back to the user when requested.

DanielLimWeiEn committed:
Add Task class to represent Tasks in the Duke bot.

Using a Task class and correspondingly Task objects to represent Tasks is a more natural way of representing tasks and allows me to keep the logic for the Task representation in a class of itself.

DanielLimWeiEn committed:
Add the ability to change the status of tasks

We can mark tasks as done and change the status of the tasks back to not done.

DanielLimWeiEn committed:
Add different types of events namely Deadlines, Events and Todos and Modify Task field's access markers

Deadline, Events and Todos are all types of Tasks and we make use to inheritance to share code. We also modify the access marker of the Task's fields so as to allow the fields to be accessible from the subclasses we created.

DanielLimWeiEn committed:
Modify String representation of Tasks

Made use of Polymorphism to modify the String representations of different tasks.

DanielLimWeiEn committed:
Add support for tracking the three new types of Tasks in Dukebot

Made use of Polymorphism to ensure correct outputs are printed to the screen.

DanielLimWeiEn committed:
Use the Input/Output redirection technique to semi-automate the testing of Duke

Add the inputs and expected outputs.

DanielLimWeiEn committed:
Teach Duke to handle errors as incorrect inputs entered by the user using Exceptions

DanielLimWeiEn committed:
Teach Duke to deal with errors associated with incorrect inputs from the user

Made use of Exceptions and Polymorphism to handle Event task, Deadline task, Todo task, Task and incorrect command exceptions.

DanielLimWeiEn committed:
Modify I/O Redirection Testing to handle new exceptions used to handle errors

DanielLimWeiEn committed:
Add support for deleting tasks from the list

DanielLimWeiEn committed:
Add test case coverage for Deletion

DanielLimWeiEn committed:
Add error handling for mark, unmark and delete commands and test coverage for mark, unmark and delete commands

DanielLimWeiEn committed:
Add JavaDoc comments for all files

DanielLimWeiEn committed:
Add File I/O Logic to read from and save Duke Tasks

DanielLimWeiEn committed:
Refactor if statement in ReadTasksFromStorage to use switch statement for cleaner code

DanielLimWeiEn committed:
Add support for Dates and Times in Duke Bot

Modify Deadline and Event to make use of LocalDateTime objects to store the Dates and Times
Change the representation of Event objects to store a start and end LocalDateTime object to use
for storing LocalDateTimes more easily and changed it's constructor. Modify Duke bot to read in
the inputs correctly.

DanielLimWeiEn committed:
Merge branch 'branch-Level-8'

DanielLimWeiEn committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/Duke.java
#      src/main/java/Event.java
#      src/main/java/Deadline.java

DanielLimWeiEn committed:
Add Exception handling for new functionality

DanielLimWeiEn committed:
Add Storage.java to handle Storage of Duke Bot

DanielLimWeiEn committed:
Add TaskList.java to handle managing the Task ArrayList

DanielLimWeiEn committed:
Add Parser.java to handle parsing user commands into a neat format for easier use

DanielLimWeiEn committed:
Add Ui.java to handle User Interface of Duke App

DanielLimWeiEn committed:
Refactor the code to extract out closely related code as classes

Add 4 classes to build in more OOP into Duke bot, namely:
* Ui : To handle user interactions
* Storage: To handle loading tasks from a file and saving tasks to the file
* Parser: To handle parsing user commands into an easier format for processing
* TaskList: To handle managing the tasks

DanielLimWeiEn committed:
Organize the classes into suitable packages

Everything is part of the duke package. The subpackages are the tasks, utilities and storage.

DanielLimWeiEn committed:
Merge commit '556af3f47a96b32898ab4cdbd65b16486a4871e8'

DanielLimWeiEn committed:
Add JUnit tests to test the behavior of the Task.java

DanielLimWeiEn committed:
Add JUnit tests to test the behavior of Todo.java

DanielLimWeiEn committed:
Add JavaDoc comments to the code.

DanielLimWeiEn committed:
Tweak the code to comply with a the module's Java Coding Standard

DanielLimWeiEn committed:
Give users a way to find a task by searching for a keyword

DanielLimWeiEn committed:
Merge branch 'branch-A-JavaDoc'

DanielLimWeiEn committed:
Merge branch 'branch-A-CodingStandard'

DanielLimWeiEn committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/task/TaskList.java
#	src/main/java/duke/utilities/Ui.java

DanielLimWeiEn committed:
Add dependencies and basic file structure to support GUi for Duke

DanielLimWeiEn committed:
Add DialogBox custom control component

DanielLimWeiEn committed:
Add MainWindow component

DanielLimWeiEn committed:
Add Main class to handle GUi logic and Add application Launcher

DanielLimWeiEn committed:
Add a GUI to Duke using the JavaFX technology

DanielLimWeiEn committed:
Add colour scheme to GUi and handle storage

DanielLimWeiEn committed:
Add GUi greeting by user and duke

DanielLimWeiEn committed:
Add error handling for GUi

DanielLimWeiEn committed:
Merge branch 'branch-Level-10'

DanielLimWeiEn committed:
Add Checkstyle configurations to be run with Gradle

DanielLimWeiEn committed:
Use checkstyle to fix coding style violations

DanielLimWeiEn committed:
Merge branch 'branch-A-CheckStyle'

april-anh commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,216 @@
+            String information = Parser.parseCommand(input);
+            String[] infoArray = information.split("\\|");
+
+            switch (infoArray[0]) {
```
Will it be clearer if you declare another variable for `infoArray[0]` to clarify the meaning of it? In this case, I guess `infoArray[0]` is a command, right?

DanielLimWeiEn replied:
That's a good suggestion 👍 . Thank you, I can update that!

april-anh commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,216 @@
+        } catch (IOException e) {
+            response = this.ui.getStringIoException(e);
+        } catch (Exception e) {
+            response = "An unexpected exception has occurred.";
```
Should you gather all the output text into separate class (maybe UI)?

DanielLimWeiEn replied:
That is a good point, I kind of hard coded that section in to handle exception, but it can be better.

april-anh commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,130 @@
+
+        File folder = new File(parts[0]);
+        if (!folder.exists()) {
+            folder.mkdir();
```
Should you name `mldir` explicitly?

DanielLimWeiEn replied:
noted

april-anh commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,130 @@
+        }
+
+        File file = new File(filePath);
+        if (!file.exists()) {
```
Will `exist()` (without **"s"**) be more consistent with the rest of your code?

DanielLimWeiEn replied:
I think exists() is the name of the method from the Java API 😅

april-anh commented on src/test/java/duke/DukeTest.java:
```diff
@@ -0,0 +1,17 @@
+    }
+
+    @Test
+    public void anotherDummyTest() {
```
Can try another meaningful name for your test or simply remove it?

DanielLimWeiEn replied:
Yup that was just a dummy test to get started using JUnit test

april-anh reviewed (commented):
Very clean and organized code :+1:

DanielLimWeiEn committed:
Modify build.gradle so that jar file created runs from proper class

DanielLimWeiEn committed:
Use assertions to document important assumptions in the code

The GUI depends on there being a running Duke instance in order to run
commands in Duke and get proper responses.

We want to make sure that the program has such a running instance other-
wise there may be unexpected or dangerous behaviour if the GUI is run.

Use Assert statements to ensure that the GUI has a functional Duke
instance for the entire program to run correctly.

Assert statements can allow the program to crash if there happens to
not be a running Duke instance which is safer than allowing the program
to continue running.

DanielLimWeiEn committed:
Refactor code and improve code quality

Several classes and methods had very long methods and made use to many
complicated expressions in the code.

This made the code less readable and more difficult to understand.

Let's refactor the code by,
* Using less complicated expressions
* Shortening methods
* Getting rid of deep nesting and arrowhead style code
* Extracting magic numbers into constants
* Making the code more explicit
* Keeping the code simple
* Improving comments for the reader

This can improve code readability and make understanding the code much
easier for both myself in the future and later users, thereby allowing
changes to be made more easily.

DanielLimWeiEn committed:
Use the Streams feature of Java in my code

DanielLimWeiEn committed:
Merge pull request #2 from DanielLimWeiEn/branch-A-CodeQuality

Refactor code and improve code quality

DanielLimWeiEn committed:
Merge branch 'master' into branch-A-Streams

DanielLimWeiEn committed:
Merge branch 'master' into branch-A-Assertions

DanielLimWeiEn committed:
Merge pull request #1 from DanielLimWeiEn/branch-A-Assertions

Use assertions to document important assumptions in the code

DanielLimWeiEn committed:
Merge branch 'master' into branch-A-Streams

DanielLimWeiEn committed:
Merge pull request #3 from DanielLimWeiEn/branch-A-Streams

Use the Streams feature of Java in my code

DanielLimWeiEn committed:
Merge branch 'master' of github.com:DanielLimWeiEn/ip

DanielLimWeiEn committed:
Set up Continuous Integration using GitHub Actions

DanielLimWeiEn committed:
Provide a way to get reminders about upcoming deadlines and events

DanielLimWeiEn committed:
Merge branch 'branch-BCD-Extension'

DanielLimWeiEn committed:
Update user guide README.md

DanielLimWeiEn committed:
Add representative screenshot and user guide

DanielLimWeiEn committed:
Add user guide

DanielLimWeiEn committed:
Update user guide

DanielLimWeiEn committed:
Add JUnit tests for Storage class

DanielLimWeiEn committed:
Add JUnit test for Deadline class

DanielLimWeiEn committed:
Add JUnit test for Event class

DanielLimWeiEn committed:
Add JUnit tests for TaskList class

DanielLimWeiEn committed:
Add JUnit tests for Parser class

DanielLimWeiEn committed:
Add JUnit tests for DukeException class

DanielLimWeiEn committed:
Add JUnit tests for Ui class

DanielLimWeiEn committed:
Touch up code quality issues by following feedback
